src/contracts/libraries/GPv2Encoding.sol

```diff
@@ -111,30 +110,53 @@ library GPv2Encoding {
         );
         order.tip = abi.decode(encodedOrder[74:], (uint256));
         uint8 flags = uint8(encodedOrder[106]);
-        order.kind = OrderKind((flags >> ORDER_KIND_BIT) & 0x1);
-        order.partiallyFillable =
-            (flags >> ORDER_PARTIALLY_FILLABLE_BIT) & 0x01 == 0x01;
         order.executedAmount = abi.decode(encodedOrder[107:], (uint256));
         uint8 v = uint8(encodedOrder[139]);
         bytes32 r = abi.decode(encodedOrder[140:], (bytes32));
         bytes32 s = abi.decode(encodedOrder[172:], (bytes32));
 
-        // NOTE: In order to avoid a memory allocation per call (which could
-        // become expensive due to the quadratic nature of memory costs), use
-        // the memory region reserved by the caller for the order result for
-        // the data to be hashed.
-        bytes32 digest;
+        order.sellToken = tokens[order.sellTokenIndex];
+        order.buyToken = tokens[order.buyTokenIndex];
+        order.kind = OrderKind((flags >> ORDER_KIND_BIT) & 0x1);
+        order.partiallyFillable =
+            (flags >> ORDER_PARTIALLY_FILLABLE_BIT) & 0x01 == 0x01;
+
+        bytes32 orderDigest;
+        bytes32 signingDigest;
+
+        // NOTE: In order to avoid a memory allocation per call by using the
+        // built-in `abi.encode`, we reuse the memory region reserved by the
+        // caller for the order result as input to compute the hash.
         // solhint-disable-next-line no-inline-assembly
         assembly {
-            mstore(order, domainSeparator)
-            // NOTE: Structs are not packed in solidity. Additionally there are
-            // 10 signed parameters per order to hash and ecrecover, for a total
-            // of `10 * sizeof(uint) = 320` bytes.
-            digest := keccak256(order, 320)
+            // NOTE: Structs are not packed in Solidity, and there are 9 order
+            // fields to hash for a total of `9 * sizeof(uint) = 288` bytes.
+            orderDigest := keccak256(order, 288)
```

Comment:
Do we invalidate orders by user? If so, why not use the nonce to track which orders have been used and avoid doing the extra hashing?

Reply:
So the main reason is we want to support two hashing schemes. For that we want the order digest to be the same regardless of the signing scheme.

> Do we invalidate orders by user?

Yes, the plan is to use `keccak(order.digest, validTo, owner)` as a key for order cancellation. In this sense, the `nonce` is just some extra order salt allowing for two orders with the same parameters and users to be disambiguated. This allows order refunds to not require the entire order data and verifying the signature.

I added some more info #14

Reply:
Basically my question would be why not use `keccak(order.nonce, validTo, owner)` and avoid the double hashing?

Reply:
Ah I see, well:
- if we use `eth_signTypedData` then we need to compute the order struct hash anyway, as spec-ed by the EIP-712.
- order storage refund encodes the order hash in the call data directly, so it does not need to be computed.
- the cost of `keccak(order.nonce, validTo, owner)` and `keccak(order.digest, validTo, owner)` is identical if we have `order.digest` computed anyway.
- `order.hash` is much harder to collide than `order.nonce`

So I think this comes back to the discussion if `nonce` should be a nonce or arbitrary data (salt, affiiliate ID, etc.). The only issue I see with using the `nonce` as an actual nonce is that it might be hard to provide a way for a user to actually determine which nonce it should use (since this is a mapping and not a list, so you can't really ready on-chain what nonce you have to use and need to keep track of it separately somehow).

Reply:
@fleupold I'm going to merge this PR as is. We can move the discussion to #14 where we can decide if we want to use an actual nonce per user, or if we want to use generic order salt (how the current misnamed `nonce` field is currently intended to work).